src/aleph/http/client.clj

```diff
@@ -487,7 +712,241 @@
                nil
                nil))))
 
+
+
+(defn- rsp-handler
+  "Returns a fn that takes a response map and returns the final Ring response map.
+
+   Shared between HTTP versions.
+   Handles errors, closing, and converts the body if not raw."
+  [{:keys [ch keep-alive? raw-stream? req response-buffer-size t0]}]
+  (fn handle-response [rsp]
+    (cond
+      (instance? Throwable rsp)
+      (d/error-deferred rsp)
+
+      (identical? ::closed rsp)
+      (d/error-deferred
+        (ex-info
+          (format "connection was closed after %.3f seconds" (/ (- (System/nanoTime) t0) 1e9))
+          {:request req}))
+
+      raw-stream?
+      rsp
+
+      :else
+      (d/chain' rsp
+                ;; chain, since conversion to InputStream may take time
+                (fn convert-body [rsp]
+                  (let [body (:body rsp)]
+                    ;; handle connection life-cycle
+                    (when (false? keep-alive?)              ; assume true if nil
+                      (if (s/stream? body)
+                        (s/on-closed body #(netty/close ch))
+                        (netty/close ch)))
+
+                    ;; Since it's not raw, convert the body to an InputStream
+                    (assoc rsp
+                           :body
+                           (bs/to-input-stream body
+                                               {:buffer-size response-buffer-size}))))))))
+
+(defn- http1-req-handler
+  "Returns a fn that takes a Ring request and returns a deferred containing a
+   Ring response.
+
+   Otherwise, puts/takes to/from the requests/responses streams."
+  [{:keys [ch keep-alive? raw-stream? requests response-buffer-size responses]}]
+  (fn [req]
+    (log/trace "http1-req-handler fired")
+    (let [resp (locking ch
+                (s/put! requests req)
+                (s/take! responses ::closed))]
+      (log/debug "http1-req-handler - response: " resp)
+      resp)))
+
+
+(defn- make-http1-req-preprocessor
+  "Returns a fn that handles a Ring req map using the HTTP/1 objects.
+
+   Used for HTTP/1, and for HTTP/2 with multipart requests (Netty HTTP/2
```

Comment:
Since we are not going to support HTTP/2 multipart requests, at least initially, I would remove the modifications that have been done to support HTTP2.
(No strong opinion)

Reply:
I think the only change is to the addition of `out-ch`.

I guess removing it is fine. I'm tempted to leave it in, since there's a decent chance the the path to HTTP/2 multipart support involves inserting `Http2StreamFrameToHttpObjectCodec` into the pipeline, at which point, the old HTTP1 code could theoretically work, but it's unclear if it's a minor thing, or will still require major adjustments.

If it still requires major adjustments to the old HTTP1 code, we might as well write a new, simpler handler that does just what we need with the HTTP1 objects.

Reply:
Removed